histogram/src/standard.rs

```diff
@@ -0,0 +1,300 @@
+    /// example, the 50th percentile (median) can be found using `50.0`.
+    ///
+    /// The results will be sorted by the percentile.
+    pub(crate) fn percentiles(&self, percentiles: &[f64]) -> Result<Vec<(f64, Bucket)>, Error> {
+        // if the histogram is empty, then we should return an error
+        if self.total_count == 0_u128 {
+            return Err(Error::Empty);
+        }
+
+        // sort the requested percentiles so we can find them in a single pass
+        let mut percentiles = percentiles.to_vec();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+        // validate all the percentiles
+        for percentile in &percentiles {
+            if !(0.0..=100.0).contains(percentile) {
+                return Err(Error::InvalidPercentile);
+            }
+        }
+
+        let mut bucket_idx = 0;
+        let mut partial_sum = self.buckets[bucket_idx] as u128;
+
+        let result: Vec<(f64, Bucket)> = percentiles
+            .iter()
+            .filter_map(|percentile| {
+                let count = (percentile / 100.0 * self.total_count as f64).ceil() as u128;
+
+                loop {
+                    // found the matching bucket index for this percentile
+                    if partial_sum >= count {
+                        return Some((
+                            *percentile,
+                            Bucket {
+                                count: self.buckets[bucket_idx],
+                                range: self.config.index_to_range(bucket_idx),
+                            },
+                        ));
+                    }
+
+                    // check if we have reached the end of the buckets
+                    if bucket_idx == (self.buckets.len() - 1) {
+                        break;
+                    }
+
+                    // otherwise, increment the bucket index, partial sum, and loop
+                    bucket_idx += 1;
+                    partial_sum += self.buckets[bucket_idx] as u128;
+                }
+
+                None
+            })
+            .collect();
+
+        Ok(result)
+    }
+
+    /// Return a single percentile from this histogram.
+    ///
+    /// The percentile should be in the inclusive range `0.0..=100.0`. For
+    /// example, the 50th percentile (median) can be found using `50.0`.
+    pub fn percentile(&self, percentile: f64) -> Result<Bucket, Error> {
+        self.percentiles(&[percentile])
+            .map(|v| v.first().unwrap().1.clone())
+    }
+
+    /// Merge the counts from the other histogram into this histogram.
+    pub fn merge(&mut self, other: &Histogram) -> Result<(), Error> {
+        if self.config.params() != other.config.params() {
+            return Err(Error::MergeIncompatibleParameters);
+        }
+
+        for (this, other) in self.buckets.iter_mut().zip(other.buckets.iter()) {
+            *this = this.wrapping_add(*other);
```

Comment:
Don't have a strong opinion on this, but do you think the semantics should be to wrap on overflow or throw an error?

Reply:
We could roll-back the change and return an error. Or we can change merge to require an allocation instead of doing in-place modification.